dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,362 @@
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_CPU_H_
+
+#include "dali/core/boundary.h"
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+/**
+ * @brief Cyclic buffer used for storing input window for convolution.
+ *
+ * Keeps `num_lanes` of each window element, intended when there is stride between
+ * window elements, but the same convolution can be applied to neighbouring innermost elements.
+ *
+ * Wraps a pointer to a buffer of type T, with size equal to length * num_lanes.
+ */
+template <typename T, int max_lanes = 16>
+class CyclicWindowWrapper {
+ public:
+  CyclicWindowWrapper(T* __restrict__ ptr, int length, int num_lanes = 1)
+      : data_(ptr), start_(0), end_(0), elements_(0), length_(length), num_lanes_(num_lanes) {
+    DALI_ENFORCE(num_lanes <= max_lanes, "Static lanes limit exceeded.");
+  }
+
+  /**
+   * @brief Drop one element consistine of `NumLanes()` lanes from the buffer, moving the start
+   *        element.
+   */
+  void PopElement() {
+    assert(elements_ > 0);
+    elements_--;
+    start_++;
+    WrapPosition(start_);
+  }
+
+  /**
+   * @brief Add a window element, that has `NumLanes()` consecutive lanes and starting at `input`,
+   * to the buffer.
+   */
+  void PushElement(const T* __restrict__ input) {
+    assert(elements_ < length_);
+    memcpy(data_ + end_ * NumLanes(), input, sizeof(T) * NumLanes());
+    elements_++;
+    end_++;
+    WrapPosition(end_);
+  }
+
+  /**
+   * @brief Add a window element to the buffer, that has `NumLanes()` lanes
+   */
+  void PushElement(span<const T> input) {
+    assert(elements_ < length_);
+    for (int c = 0; c < NumLanes(); c++) {
+      data_[end_ * NumLanes() + c] = input[c];
+    }
+    elements_++;
+    end_++;
+    WrapPosition(end_);
+  }
+
+  /**
+   * @brief Get pointer to the start of pixel `idx` in the buffer, taking into account the cyclic
+   * wrapping
+   */
+  T* GetElementOffset(int idx) const {
+    assert(idx < elements_);
+    if (start_ + idx < length_) {
+      return data_ + (start_ + idx) * NumLanes();
+    } else {
+      return data_ + (start_ + idx - length_) * NumLanes();
+    }
+  }
+
+  /**
+   * @brief Calculate dot product with 1-channel `window` length equal to the one specified
+   * at construction. The result is stored in pixel stored at `accum`.
+   */
+  template <typename W>
+  void CalculateDot(W* __restrict__ accum, const W* __restrict__ window) const {
+    assert(elements_ == length_);
+    for (int c = 0; c < NumLanes(); c++) {
+      accum[c] = 0;
+    }
+    int window_idx = 0;
+    for (int buf_idx = start_; buf_idx < length_; buf_idx++, window_idx++) {
+      for (int c = 0; c < NumLanes(); c++) {
+        accum[c] += window[window_idx] * data_[buf_idx * NumLanes() + c];
```

Comment:
nitpick:
`window[window_idx]` and `data_[buf_idx * NumLanes()]` could be calculated outside of the loop